Recognize where clauses

nikkirad opened the pull request:
Modified the regular expression `chpl_sig_pattern` and added functionality to recognize where clauses.

nikkirad committed:
Added functionality to recognize where clauses.

nikkirad committed:
Edited line lengths

nikkirad committed:
Edited line (again)

nikkirad committed:
Edit and add tests for where clause

nikkirad committed:
Written in Chapel

nikkirad committed:
Added missing comma

lydia-duncan commented on sphinxcontrib/chapeldomain/__init__.py:
```diff
@@ -35,7 +35,7 @@
 from sphinxcontrib.chapeldomain.chapel import ChapelLexer
 
-VERSION = '0.0.25'
+VERSION = '0.0.26'
```
This is already covered by Ahmad's open PR - @arezaii, why don't you go ahead and merge that PR and Nikki can rebase?

lydia-duncan commented on test/test_chapeldomain.py:
```diff
@@ -766,51 +767,56 @@ def test_with_prefixes(self):
         ]
         for sig, prefix, name, arglist in test_cases:
-            self.check_sig(sig, prefix, None, name, arglist, None)
+            self.check_sig(sig, prefix, None, name, arglist, None, None)
```
We should add a test specifically for where clauses only, like the other aspects have.  It's good that you've added them to the `test_with_all` test, but there should be a `test_with_where_clause` or something and it should show more styles of where clauses than just `where a > b`

lydia-duncan commented on test/test_chapeldomain.py:
```diff
@@ -766,51 +767,56 @@ def test_with_prefixes(self):
+            ('iter foo() ref', 'iter ', None, 'foo', '', ' ref', None),
+            ('inline proc Vector.pop() ref', 'inline proc ', 'Vector.', 'pop', '', ' ref', None),
+            ('inline proc range.first', 'inline proc ', 'range.', 'first', None, None, None),
+            ('proc My.Mod.With.Deep.NameSpace.1.2.3.432.foo()', 'proc ', 'My.Mod.With.Deep.NameSpace.1.2.3.432.', 'foo', '', None, None),
```
Looks like you dropped the Math.fib check that would go above this line, was that intentional?

lydia-duncan commented on test/test_chapeldomain.py:
```diff
@@ -766,51 +767,56 @@ def test_with_prefixes(self):
-            ('record MyRec:SuRec', 'record ', None, 'MyRec', None, ':SuRec'),
-            ('record N.R : X, Y, Z', 'record ', 'N.', 'R', None, ': X, Y, Z'),
+             'proc ', None, 'RandomStream', 'seed: int(64) = SeedGenerator.currentTime, param parSafe: bool = true', None, None),
+            ('class X', 'class ', None, 'X', None, None, None),
```
I don't think this is your fault, but this seems like a weird spot to be checking class and record signatures, and we definitely shouldn't allow the possibility of class or record signatures including where clauses - if there's a way to add a check to ensure that, that'd be a good thing to do.

nikkirad replied:
Hi Lydia, not sure what you mean by this. Do you mind elaborating?

lydia-duncan replied:
Sure!  Most of the checks here are on functions - procs, iters, operators.  Those all follow a set pattern that looks pretty similar.  Record and class signatures are a different sort of declaration, though I suppose the difference is more distinct in Chapel code than in the generated Sphinx signatures.  But because where clauses are only ever associated with functions, it definitely wouldn't make sense for a Sphinx record or class signature to include it, so it seems odd to me for them to even have a where clause entry.  The value in that case should always be None.

If we can't make it so that record and class signatures avoid having that argument in the group, it seems like it would be nice to have a check somewhere in the __init__.py file that will raise a ValueError or something if we know we've been given a record or class and the where_clause is set to something other than None.

Does that help clarify?  I'm happy to have a meeting to explain it better, too

lydia-duncan commented on test/test_chapeldomain.py:
```diff
@@ -766,51 +767,56 @@ def test_with_prefixes(self):
+            ('operator string.+(s0: string, s1: string) : string', 'operator ', 'string.', '+', 's0: string, s1: string', ' : string', None),
+            ('operator *(s: string, n: integral) : string', 'operator ', None, '*', 's: string, n: integral', ' : string', None),
+            ('inline operator string.==(param s0: string, param s1: string) param', 'inline operator ', 'string.', '==', 'param s0: string, param s1: string', ' param', None),
+            ('operator bytes.=(ref lhs: bytes, rhs: bytes) : void ', 'operator ', 'bytes.', '=', 'ref lhs: bytes, rhs: bytes', ' : void ', None),
```
I believe operators can also have where clauses, so it would be good to add a test including one.

lydia-duncan reviewed (approved):
Nice!  This looked tricky, just a few suggestions and questions

nikkirad committed:
Created test_with_where_clause

nikkirad committed:
Proc-like Check for Where Clauses

Raise a ValueError if we a non-proc-like directive contains a where_clause that is set to something other than None.

lydia-duncan reviewed (approved):
Awesome, thanks!
